pkg/kubernetes/scripts/virtual-machines/components/createVmButton.jsx

```diff
@@ -0,0 +1,112 @@
+
+import { connect } from 'react-redux'
+import React, { PropTypes } from "react";
+import DialogPattern from 'cockpit-components-dialog.jsx';
+import CreateVmDialog from './createVmDialog.jsx'
+
+import { mouseClick } from '../utils.jsx'
+
+const _ = cockpit.gettext;
+
+
+class CreateVmButton extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.createVmDialog = this.createVmDialog.bind(this);
+        this.createVm = this.createVm.bind(this);
+    }
+
+
+    createVm(resource) {
+        const deferred = cockpit.defer();
+        const promise = this.props.kubeMethods.create(resource, this.props.kubeLoader.limits.namespace);
+
+        // wrap angular promise in cockpit promise because of incompatibility
+        promise.then(function(/* ... */) {
+            return deferred.resolve.apply(deferred, arguments);
+        });
+
+        promise.catch(function(result) {
+            deferred.reject(`${result.status}: ${result.message}`);
+        });
+
+        return deferred.promise;
+    }
+
+    createVmDialog() {
+        let getResult;
+        let validate;
+
+        const dialogBody = (
+            <CreateVmDialog setValidateCallback={(callback) => validate = callback}
+                            setGetResultCallback={(callback) => getResult = callback}/>
+        );
+
+        const dialogProps = {
+            title: _("Create New Virtual Machine"),
+            body: dialogBody,
+        };
+
+        // also test modifying properties in subsequent render calls
+        const footerProps = {
+            actions: [
+                {
+                    clicked: () => {
+                        if (typeof validate !== 'function' || typeof getResult !== 'function') {
+                            console.error("Uninitialized returnValues!");
+                            return cockpit.defer().reject(_("UI Internal Error: see console log.")).promise;
+                        }
+
+                        if (validate()) {
+                            const result = getResult();
+                            return this.createVm(result.resource);
+                        } else {
+                            return cockpit.defer().reject(_("Resolve above errors to continue")).promise;
+                        }
+                    },
+                    caption: _("Create"),
+                    style: 'primary',
+                },
+            ],
+        };
+
+        DialogPattern.show_modal_dialog(dialogProps, footerProps);
+    }
+
+    render() {
+        return (
+            <a className="card-pf-link-with-icon pull-right" id="create-new-vm"
+               onClick={mouseClick(() => this.createVmDialog())}>
+                <span className="pficon pficon-add-circle-o"/>{_("Create")}
+            </a>
+        );
+    }
+}
+
+CreateVmButton.propTypes = {
+    kubeLoader: PropTypes.object.isRequired,
+    kubeMethods: PropTypes.object.isRequired,
+}
+
+export default connect()(CreateVmButton)
```

Comment:
I miss the reason for that here.

Reply:
I just left it there because I used it before